### PR TITLE
Print logs to stderr

### PR DESCRIPTION
**Pull request by phanimarupaka**
This PR is to make printer print logs to stderr instead of stdout. Only resources will the printed to stdout.

### CODE REVIEW

File: e2e/testdata/fn-eval/missing-fn-config/.expected/config.yaml
Comment (frankfarzan) [inline e2e/testdata/fn-eval/missing-fn-config/.expected/config.yaml:20]: So this is not part of stderr?

File: e2e/testdata/fn-render/missing-fn-image/.expected/config.yaml
Comment (frankfarzan) [inline e2e/testdata/fn-render/missing-fn-image/.expected/config.yaml:17]: Why is this not part of stdErr?

**Review by frankfarzan — DISMISSED** [review]

File: e2e/testdata/fn-eval/missing-fn-config/.expected/config.yaml
Comment (phanimarupaka) [reply e2e/testdata/fn-eval/missing-fn-config/.expected/config.yaml:20]: This will be part of stderr, but we do only the substring match here. We only check for the message relevant to the test.

There are other tests which check of this is included in stderr. We need not check that in all the tests.

File: e2e/testdata/fn-render/missing-fn-image/.expected/config.yaml
Comment (phanimarupaka) [reply e2e/testdata/fn-render/missing-fn-image/.expected/config.yaml:17]: Same as above.

File: e2e/testdata/fn-eval/missing-fn-config/.expected/config.yaml
Comment (frankfarzan) [reply e2e/testdata/fn-eval/missing-fn-config/.expected/config.yaml:20]: Link to tests that assert the non-structure output for:

- eval
- eval | eval
- source | eval
- render
- render | eval

Comment (frankfarzan) [reply e2e/testdata/fn-eval/missing-fn-config/.expected/config.yaml:20]: @droot @Shell32-Natsu

File: internal/util/printerutil/printerutil.go
Comment (droot) [inline internal/util/printerutil/printerutil.go:26]: If printer supports only writing to one stream, this option becomes redundant.

File: internal/printer/printer.go
Comment (phanimarupaka) [reply internal/printer/printer.go:75]: @droot Yes. Currently it is only being used by fn commands. All live commands use live printer different from this. I removed it for now as we don't need it and when live also uses this printer we can introduce the stdout option. For now, to be clear, the printer only prints to stderr and this code reflects that.

File: e2e/testdata/fn-eval/missing-fn-config/.expected/config.yaml
Comment (phanimarupaka) [reply e2e/testdata/fn-eval/missing-fn-config/.expected/config.yaml:20]: @frankfarzan Added third commit which covers all these cases.

File: internal/util/printerutil/printerutil.go
Comment (phanimarupaka) [reply internal/util/printerutil/printerutil.go:26]: There is no option there. It is only writing to stderr stream.

File: e2e/testdata/fn-render/out-of-place-fnchain-stdout-results/.expected/config.yaml
Comment (frankfarzan) [inline e2e/testdata/fn-render/out-of-place-fnchain-stdout-results/.expected/config.yaml:1]: I think we should omit beginning of output like this, otherwise, it's not really testing what we want.

This is not the expected sttderror from:

```
kpt fn render -o stdout --results-dir $results \| kpt fn eval - --image gcr.io/kpt-fn/set-annotations:v0.1.3 --results-dir $results -- foo=bar \| kpt fn eval - --image gcr.io/kpt-fn/set-labels:v0.1.3 --results-dir $results -- tier=backend
```

Comment (phanimarupaka) [reply e2e/testdata/fn-render/out-of-place-fnchain-stdout-results/.expected/config.yaml:1]: This is to test if the substring we expect is present in stderr and not checking the entire stderr. Not sure if there is a better way to refer the temporary directory created in exec.sh.

Comment (phanimarupaka) [reply e2e/testdata/fn-render/out-of-place-fnchain-stdout-results/.expected/config.yaml]: Updated stderr. But it will include the progress of render only.

Comment (droot) [reply e2e/testdata/fn-render/out-of-place-fnchain-stdout-results/.expected/config.yaml]: This is little tricky because in each of the command `structured-result-file` will depend on the temporary directory, so you can do only partial assert here and in this case, substring match only for the first command. 

test harness can be extended to support make temporary dir etc. available in a templatized form, but not sure if we want to go there.

But we should be able to test `piped` version without `structured-result` for non-structured CLI output. So, lets have a version of this test without structured result just for testing the presentation output. That should be sufficient.

File: internal/printer/printer.go
Comment (mortent) [reply internal/printer/printer.go:75]: This is not correct. The printer is used by the pkg commands as well as some of the live commands.
Does this lead to all print statements going to stderr? That doesn't seem right.

**Review by mortent — CHANGES_REQUESTED** [review]

File: e2e/testdata/fn-render/out-of-place-fnchain-stdout-results/.expected/config.yaml
Comment (phanimarupaka) [reply e2e/testdata/fn-render/out-of-place-fnchain-stdout-results/.expected/config.yaml]: That is covered as part of other tests. This test is to test the behavior when the resources are piped and results are written to --results-dir. I updated the test, it will test the render result which is good enough IMO. Please look at updated file.

File: e2e/live/end-to-end-test.sh
Comment (mortent) [inline e2e/live/end-to-end-test.sh:484]: With this change, it seems we end up with live commands sending output to a combination of stdout and stderr. This should be fixed consistently across the full kpt UX.

File: internal/printer/printer.go
Comment (mortent) [inline internal/printer/printer.go:100]: Similar to my comment above, this is being used by packaging. We should make sure this is changed consistently across all kpt commands.

**Review by mortent — CHANGES_REQUESTED** [review]

File: internal/printer/printer.go
Comment (phanimarupaka) [reply internal/printer/printer.go:75]: Why not ? The intention is to print the diagnostic information to stderr and actual resources to stdout so that they can be piped to other commands/tools. That makes clear distinction of the output locations.

@frankfarzan @droot

File: internal/printer/printer.go
Comment (phanimarupaka) [reply internal/printer/printer.go:100]: So, the package results logs are also printed to stderr. What more changes should be done for pkg group ?

File: internal/printer/printer.go
Comment (mortent) [reply internal/printer/printer.go:75]: The posix guidance for what should go to stdout vs stdin is somewhat vague, but I can see the argument for printing everything to stderr. We should just make sure we do it consistently across all of kpt.

File: e2e/live/end-to-end-test.sh
Comment (phanimarupaka) [reply e2e/live/end-to-end-test.sh:484]: Does setting [this](https://github.com/GoogleContainerTools/kpt/blob/master/commands/livecmd.go#L57) to os.StdErr is good enough to direct all logs to stderr ?

Comment (mortent) [reply e2e/live/end-to-end-test.sh:484]: No, the IOStreams is not used consistently across all commands. I actually think we should try to phase out the use of IOStreams and instead rely on getting the input/output streams from cobra. But I haven't had time to fix that yet.

Comment (phanimarupaka) [reply e2e/live/end-to-end-test.sh:484]: Can we have fixing live commands output as a follow up item ? We need this in for fn commands which is of high priority.

File: internal/printer/printer.go
Comment (phanimarupaka) [reply internal/printer/printer.go:75]: I agree that we should have that consistency but not sure if we should do it as part of single PR.

File: internal/printer/printer.go
Comment (mortent) [reply internal/printer/printer.go:100]: I think the printing is mostly done through the printer for packaging. I think the exception might be the output from git that is piped to the output. It might already end up on stderr, but I'm not sure.

File: e2e/live/end-to-end-test.sh
Comment (mortent) [reply e2e/live/end-to-end-test.sh:484]: Yeah, we can create a separate issue for m3 for fixing this for live.

File: internal/printer/printer.go
Comment (droot) [reply internal/printer/printer.go:75]: So far what we print in `fn` commands is purely progress/diagnostic information, which is sort-of okay to print on `stderr` (POSIX guidelines are not super clear there) and that gives us consistency in `fn` command output in piped/non-piped mode. But for commands such as `kpt live status`, I would expect the output to be printed on `stdout` instead of `stderr`.

I would have liked this printer to have the ability to be able to print on `stderr` and `stdout`, so that this can abstract away printing to CLI from all `kpt` commands. Not having support for `stdout` will make `fmt.Printf` attractive.

Comment (mortent) [reply internal/printer/printer.go:75]: I agree that for some commands (in particular the live commands), it becomes a bit harder to suggest everything should be printed to stderr. I commented on https://github.com/GoogleContainerTools/kpt/issues/2178, but this seems very focused on the `fn` commands without considering the full kpt UX. Regardless of whether each command is meant to be composed, we should have a consistent approach to what is printed to stdout vs stderr.

Another option would be to have kpt check if the command is being used with pipes and then change what is being printed and whether we use stdout/stderr based on that. The drawback here is that this might be considered confusing.

File: e2e/live/end-to-end-test.sh
Comment (phanimarupaka) [reply e2e/live/end-to-end-test.sh:484]: https://github.com/GoogleContainerTools/kpt/issues/2191

File: internal/printer/printer.go
Comment (droot) [reply internal/printer/printer.go:75]: You are right about having consistent approach regardless of the composability.

I found this post very relevant https://mehulkar.com/blog/2017/11/stdout-vs-stderr/

> I’ve always thought that stderr for a program is to display errors and stdout is for “everything else”. But it turns out stdout is meant to be program output and stderr is meant to be… “everything else”. This was conceptually difficult for me until I thought of an example.
> 
> For example, in a calculator program, if I input 1 + 1, the output is 2. This should be printed to stdout. However, if there is internal logging, that should be printed to stderr. E.g. “Gathering input”, “initializing adding machine”, etc.
> 

So we can apply this principle to all the kpt commands. So, for every command, first question to ask is `what is the output` of the command and everything else goes to the `stderr`. WDYT ?

Comment (phanimarupaka) [reply internal/printer/printer.go:75]: I agree with this. This has clear definition.

File: e2e/testdata/fn-render/fnresult-fn-success/.expected/config.yaml
Comment (frankfarzan) [inline e2e/testdata/fn-render/fnresult-fn-success/.expected/config.yaml:21]: @droot Why do we have idempotent packages? We should remove this option all-together.

Comment (frankfarzan) [reply e2e/testdata/fn-render/fnresult-fn-success/.expected/config.yaml:21]: https://github.com/GoogleContainerTools/kpt/issues/2193

**Comment by frankfarzan** [comment]
lgtm for tests and fn UX. I'll let others approve.

File: e2e/testdata/fn-render/fnresult-fn-success/.expected/config.yaml
Comment (droot) [reply e2e/testdata/fn-render/fnresult-fn-success/.expected/config.yaml:21]: Good question. This test-case tests function result for success case and uses `search-replace` function. Now `search-replace` output for first run is different from subsequent run because once the replacement is done, no output.

If we have a function that always produces same structured result, we don't need this test to be non-idempotent. Probably, it's easier to write a test or demo function to achieve that.  I believe this is the only test-case that runs into this problem.

Comment (frankfarzan) [reply e2e/testdata/fn-render/fnresult-fn-success/.expected/config.yaml:21]: Idempotency only applies on changes made to the filesystem. `search-replace` should be idempotent in that regard. You may want to relax what the test harness asserts.

File: internal/cmdinit/cmdinit_test.go
Comment (mortent) [inline internal/cmdinit/cmdinit_test.go:173]: What is this testing? It seems like the buffers aren't being used, so they can never contain anything else than `""`.

File: internal/printer/fake/fake.go
Comment (mortent) [reply internal/printer/fake/fake.go:41]: Is there a way we can avoid requiring almost all users to end up using `CtxWithFakePrinter(nil, nil)` here? It just creates noise. Maybe provide a second function that takes the two writers that can be used by the few use-cases where it is needed.

File: internal/printer/printer.go
Comment (mortent) [inline internal/printer/printer.go:38]: I think we need a way for users to fetch the ErrStream too here. An example is this:
https://github.com/GoogleContainerTools/kpt/blob/fc64a0a358302ec41b3184188334e57fec778b1d/internal/gitutil/gitutil.go#L107
I assume we are doing something similar when piping the output from docker commands to the console.

**Review by droot — COMMENTED** [review]
LGTM. You can merge it after @mortent approval.

File: internal/util/printerutil/printerutil.go
Comment (phanimarupaka) [reply internal/util/printerutil/printerutil.go]: Updated.

File: internal/fnruntime/runner.go
Comment (phanimarupaka) [reply internal/fnruntime/runner.go]: Done.

File: internal/printer/fake/fake.go
Comment (phanimarupaka) [reply internal/printer/fake/fake.go]: Updated.

File: internal/util/printerutil/printerutil.go
Comment (phanimarupaka) [reply internal/util/printerutil/printerutil.go]: Done.

File: internal/cmdinit/cmdinit_test.go
Comment (phanimarupaka) [reply internal/cmdinit/cmdinit_test.go]: Removed.

File: internal/printer/printer.go
Comment (phanimarupaka) [reply internal/printer/printer.go:38]: Updated.

File: internal/printer/fake/fake.go
Comment (mortent) [inline internal/printer/fake/fake.go:45]: We should probably use `CtxWithPrinter(io.Discard, io.Discard)` here instead of passing in nil here. Using nil will cause panic if the code tries to use either of the writers.

File: internal/printer/fake/fake.go
Comment (mortent) [inline internal/printer/fake/fake.go:50]: I usually prefer to just discard output from tests instead of sending it to the console. It just tends to make the test output hard to read.

**Review by mortent — APPROVED** [review]
Two small comments, but overall looks good.

**Review by droot — APPROVED** [review]

**Pull request merged**
